### ISSUE TITLE
Detect if a user is attempting to use Proton on a NTFS drive (Non-EXT4)

### ISSUE THREAD

**Issue by ruineka** [ValveSoftware/steam-for-linux#8274]
I've been helping new users use Proton and I pretty regularly find people wasting their time and datacaps to only find out the hard way that they are not able to play games using a NTFS formatted drive. I think it would be wise for Steam to detect what drive a user is attempting to launch a game with using SteamPlay saying (Launching a game on a non EXT4 drive is almost guaranteed to fail, Try anyway?). As far as I know their is no indication anywhere within steam telling the user not to use NTFS formatted drives with SteamPlay and when a game fails to launch the user gets frustrated.

**Comment by smcv** [comment]
Proton on NTFS drives can be made to work, but it's less robust than on Unixish filesystems like ext4, xfs or btrfs, and it needs specific mount options. More information here: https://github.com/ValveSoftware/steam-runtime/issues/434#issuecomment-885605450

**Comment by insanemal** [comment]
@ruineka It doesn't have to be ext4. Most posix complaint filesystems are fine. I use XFS on all of my devices without issue. 

Your "Non ext4" statement is patently false

**Comment by ruineka** [comment]
> @ruineka It doesn't have to be ext4. Most posix complaint filesystems are fine. I use XFS on all of my devices without issue. 
> 
> Your "Non ext4" statement is patently false 

I've never used anything other than EXT4, Fat, and NTFS. I don't know anything about other filesystems. The NTFS not working properly with Proton without tinkering still is an issue though and many people new people try Linux, download steam and run games from a NTFS drive and the result is clicking play..it shows "running" only to show play again.

**Comment by ruineka** [comment]
Oh wow, judging by this Valve already made a prompt addressing this issue https://www.reddit.com/r/linux_gaming/comments/rwggd8/im_getting_an_error_called_new_steam_library_must/?utm_medium=android_app&utm_source=share

**Comment by Magi-KS** [comment]
side note, when used with the NTFS kernel module in latest linux(5.15), it freezes my system and i have to use `chkdsk /F` on it if I want to mount it again in linux.

**Comment by Mitsunee** [comment]
During my years on linux so far I've had three instances of steam automatically installing Proton and Runtime Soldier on my old NTFS HDD that I keep my games on. I have always had my EXT4 drive set as the default install location and that was ignored every time. Not sure how this is still a thing that can happen given that it's always been known that steam's runtimes do not work properly on NTFS.

**Comment by Critified** [comment]
Good evening, 

Many users i guess like myself tried to use huge pre-existing Steamlibrary drive to save data and disk storage. 
I had over 4TB of games on a explicit Steamlibrary drive which I tried to use under linux, not only it does not start games, 
it actualy physicaly broke the rest of the drive, it went permanently read only as far as root folder, with no possibility to delete anything at all and full format was required. It got the data that shouldnt be touched by steam write protected. 

Personally I thought some sort of rights management issue was causing this, so I tried to format it to only existing permission-less file system both Ms and Unixes sort of 'support' which is exfat but that got me into even more trouble. 

I would suggest that steam should upon detecting ntfs/fat type partition at the very least make a very very very clear warning to user that using it may make all existing data on the partition unreadable, as this partition type has limitations.
Long term suggestion would be to fix it and make sure people can use NTFS drives with no user input required. Theres going to be many many people that do not want to redownload hundreds of gigs of games. This will only lead to more frustration. Was I only a bit less patient or experienced I would have wiped pinguin already.

**Comment by Mitsunee** [comment]
> Good evening,
> 
> Many users i guess like myself tried to use huge pre-existing Steamlibrary drive to save data and disk storage. I had over 4TB of games on a explicit Steamlibrary drive which I tried to use under linux, not only it does not start games, it actualy physicaly broke the rest of the drive, it went permanently read only as far as root folder, with no possibility to delete anything at all and full format was required. It got the data that shouldnt be touched by steam write protected.
> 
> Personally I thought some sort of rights management issue was causing this, so I tried to format it to only existing permission-less file system both Ms and Unixes sort of 'support' which is exfat but that got me into even more trouble.
> 
> [...]

@Critified First of all, are you sure the drive is mounted correctly? NTFS drives appearing as readonly is a very common figuration issue with the NTFS drivers on linux. For reference my HDD is mounted with the options `ntfs  auto,nofail,noatime,rw,user,exec 0 0` in my fstab file.

I also don't exactly understand how you went from "write protected" to "unreadable". I'd also like to note that SSDs in particular are supposed to go into read-only mode if they detect the end of their lifespan approaching to make sure the user can recover their important files before the flash storage dies.

-----

As for the actual topic of this issue: Libraries on NTFS mostly work fine, you'd expect as much for windows games on a windows filesystem at the very least. What is causing problems is when steam installs proton and WINE prefixes to an NTFS drive as described in my comment above. One trick I have found is setting up a symlink between your drives like this:

```sh
ln -s $HOME/.steam/steam/steamapps/compatdata /path/to/your/drive/steam/steamapps/compatdata
```

This makes sure steam returns to your non-NTFS drive (in this case my EXT4 system SSD) to install wine prefixes there. As for proton you would have to uninstall and manually reinstall it and the other steam runtimes to your non-NTFS drive.

TL;DR: Steam does seemingly no attempt to detect the filesystem of a location and should automatically fall back to a non-NTFS drive when installing runtimes and compatdata. It also still ignores the default library when automatically installing runtimes as a dependency for running a game from another drive.

**Comment by insanemal** [comment]
Did you remember to disable fast shutdown in windows?

I'm betting you didn't, which would explain mangling both NTFS and exfat
file systems.

Also, I don't understand why you're getting snooty this ticket is a
dumpster fire. You're missing details, you have a poorly remembered, and
even more poorly described situation that has issues that don't match
standard behaviour, yet have no logs or other useful sources of information
that might allow us to make sense of your ramblings.

You're trying to do something that is on the edge of what is supportable,
with next to no factual information, just rough outlines, all the while
acting like it's everybody else's fault we can't figure out what you
actually did.

Sounds like a great way to get help.

Oh and you're trying to request a feature that isn't required by anybody
else. (And probably isn't required by you but you can't properly explain
wtf you did so nobody can be sure)

I'm betting you mounted the NTFS volume as your damn home folder too.

Stop being a jackass.

**Comment by Mitsunee** [comment]
\*sigh\* and here I thought the constant fighting and unprofessional nonsense would be over after I unsubscribed from all the tickets on microsoft products. Makes me question my intention to get involved with open source more. /unsubscribe

**Comment by insanemal** [comment]
What are you talking about?

Also, since when is steam open source?

**Comment by insanemal** [comment]
Also on the suggestion to "just detect NTFS and put it somewhere else" which I mean sounds great, but where else can you put things?

If the user has done something odd like mounted up an NTFS drive as their home folder  none of the standard locations for dot folders are usable. So where else does a non-root user have, by default, sufficient permissions to create files/folders that isn't /tmp? (which should have noexec, so also useless) 

It can't be a one size fits all message or it's not really that much more useful than no message.  But for it to be actually useful it would need to work out the users config and then figure out if that even allows for a working config. Which will end up with a lot of very fragile logic that only covers a few "sensible" configs. Not to mention all the fun things Microsoft do that plays very fast and loose with on disk consistency. (I'm looking at you fast shutdown). Should Steam be advising people about how to adjust Windows settings as well? (I mean it's too late by that point, the damage is potentially done)   

How hard should Valve work at trying to make a very dangerous (to your data) config "Just Work" when doing so makes not losing user data their responsibility?

**Comment by ruthan** [comment]
I have this bug too simulated on multiple machines, with recent Debian / Ubuntu issue.. When i run game from ntfs partition, which i using as shared for multiple OS.. i get proton crash, when i move it to to EXT4 its working..

Log is full these messages:
`src/clientcommon/clienthelpers.cpp (27) : Assertion Failed: FixupSteamAppsPathIfNecessary: unexpected 'SteamApps' "/home/ruthan/.local/share/Steam/steamapps/common/Proton - Experimental" -> "/home/ruthan/.local/share/Steam/SteamApps/common/Proton - Experimental"`

When i try to use some not root ntfs folder for Steam library i get crash.. Im using ntfs3g..

**Comment by Xarishark** [comment]
+1 on this feature. I provide support on the Bazzite discord and I have to specify with users if they are playing their games from their NTFS partition. Please include a notification at the very least. This should be implemented asap.

**Comment by ruthan** [comment]
Its 4 years when it was reported!

I read something like that Linux games really on some permission which are not available on NTFS, but on Windows its running fine, so it should be possible just disable some features which require these permission to make it working.

  I hit other Linux specific issues, which just work on Windows, for example sharing games library for multiple users - again on WIndows it just works, on Linux you hit permission tell..  It seems to be another Linux/Unix Mainframe heritage which only complicate basic user life, normal user does not need server like permissions framework..
   - I had to fight with terminal set new user, set some user groups, set some file mask, set some rules for new files to use mask, set some checking scripts to make it working.. and still there was like issue that set /var library, instead /var/home as default on Bazzite, not everytime work and often it was back to /var/home

   I had problem with MangoHud exec script values reading when it run through Wine / Proton too, again because of permission hell. 

 Yeah Linux is getting better, but lots of legacy bad stuff still keeps happening and if you wait 4 years for fixes, something is very wrong..

**Comment by stellaberrant** [comment]
+1,  please at least give a warning when someone is trying to add a NTFS drive as a Steam Library, it would mean less headache for Valve (less bug reports), less headache for users, less headache for distributions. It's a win-win situation for everyone. 

Additionally, I know someone did a long term experiment on using NTFS for a Steam Library, and the results were… not good. All of the drives had data corruption after a few months, one had lost several months of data, 1 needed to be wiped.

**Comment by insanemal** [comment]
At a bare minimum you need to put out a statement that it's not recommended or supported. There are people on Reddit telling people it's fine and it works and they are doing it without disabling fast shutdown and all kinds of crazy stuff.

**Comment by ruthan** [comment]
They will always find some excuse, loophole as its possible but not supported, you are on your own.. i call these things Linux "quality".

  Ntfs is great / best for sharing data for multiboot machines, its supported by Linux so software should be aware of this.. exFat is much less safe and i would expected that you will hit similar issues, but its even less capable than NTFS..

   And exFat has its own issues, there is not exFat to other system data safe convertor and its actually much worse from point of data recovery softwares, i did it multiple times - NTFS or FAT32 are much less headache, so not recommend exFat.  

 Funny thing about Wine / Proton is there is more new bugs, than fixed bugs.. such products simply sucks. It got so bad, that you need specific version for specific game and what worked with older version, its now broken. Yeah not great.

   Which good product would not handle so major bug in there product for 4 years? Its complicated so as developers are most volunteers, they are often ignoring major hard to solve bugs.. 

   So +1 for year of Linux desktop became good enough, its just bad design, but code, bad support, bad OS as service for user.. 

  Yeah it has some great features like KVM card passthrough, multiuser support without user hacks etc.. but list of negatives are much longer. It just works is often not a thing.

**Comment by insanemal** [comment]
Literally none of what you just said is accurate or helpful. 

NTFS is a terrible filesystem to share files with because windows doesn't properly close the filesystem unless you adjust its default settings. And many people do not do this. This is because Microsoft don't want or need to ensure your drives work outside of Windows. 

Everything else you've said is wrong. Like doesn't need to be addressed wrong.

**Comment by ruthan** [comment]
Im describing real life facts, im not helpful? Who is not helpful are developers which arent able to fix 4 years bug and ignoring it.. other mentioned bugs. I give you other analogies, how long we are waiting on Wayland to just work, is supposed to almost done for 8 years or so.

 Here we have what i wrote, blame anything but LInux in realtime. NTFS is most used filesystem, you cant ignore it and claim that you product ready for general usage, not only for multibooting, but because of external devices.

 Closing filesystem, what it suppose to be? Do you mean that annoying thing when Linux refuse to work with NTFS volume because it has some dirty flag, which Linux is actually able to fix by one command in most of cases?

  Good analogy is MacOS, there you can get NTFS support by 3rd party software and it works fine. Paragon actually tried to make it for Linux too for free, but they hit our kernel dictator and there were some discussions, back and forth and last i checked there still some supposed files corruption, but nobody is able to fix it, at least AI still claims that faster/newer NTFS implementation is still not safe. 
   In reality problem is caused by excessive permissions (Linux is already able to bypass them for NTFS) and normal user can care less about some files permission on his game library. There nothing to steal, even i would be freak about my game saves and settings security, its already in ~/home.. there nothing to protect in Steam games library, there just binaries and assets. Especially if there is already Steam inbuild DRM protection, so if someone would steal data, he still need own games under their account.

   Im not helpful, because im criticize something. Well im doing it because  present state of the product is not worthy of praise.. Point out problem, is always first step to fix it, but we are playing blame games. Im using LInux for ~25 years, year of desktop linux is still not here.

   BTW other Valve problem, SteamOS was at the start was for all PC machines, which met minimal specs, after they changed it to Steam Deck only and promised make it for all PCs soon, how long we are waiting 2 years?

**Comment by insanemal** [comment]
Again, nothing you have said here is correct or helpful. There are some tiny nuggets of fact, none relevant to the discussion however. 

You seem to be missing the biggest fact, why I am unsure. That fact being that most of the NTFS corruption issues are thanks to Windows and it's default behaviour that many people don't understand how/why to change. 

This is not something that can be dealt with outside of windows and isn't a failing of Linux or any of its NTFS drivers as Mac users will experience the exact same thing with NTFS on "non-removable" drives. 

@kisak-valve or someone probably needs to do some moderation here.

**Comment by Xarishark** [comment]
Please lets stay on topic and keep things helpful. There is nothing to be gained here from being toxic. 

Back to the problem at hand. We had to implement a workaround on Bazzite to nag the user every time that they mounted a NTFS partition to remember that they cant play games from it. Its crude and we would love for Valve to implement something in their code that does this properly IN UI. Please. Just **block** or **notify** the user that NTFS partitions are not supported and should never be used for gaming. Please
